libraries/scopes/src/expressions.js

```diff
@@ -36,6 +37,64 @@ exports.satisfiesExpression = function(scopeset, expression) {
   return isSatisfied(expression);
 };
 
+/**
+
+ * Given a scope expression and set of scopes that satisfies it, return a
+ * subset of those scopes that satisfies the scope expression.  This subset is
+ * not necessarily minimal: in the case of AnyOf, it includes all scopes
+ * satsifying any branch of that alternative.  To do otherwise would return
+ * only one of several possible minimal scope sets.
+ *
+ * Returns undefined if the scopeset does not satisfy the expression.
+ */
+exports.scopesSatisfying = (scopeset, expression) => {
+  let scopes = [];
+  /* Evaluate the given expression, appending all used scopes to `scopes` and
+   * returning true if satisfied, otherwise returning false and leaving `scopes`
+   * as it was found.
+   */
+  const recurse = expr => {
+    if (typeof expr === 'string') {
+      if (scopeset.some(s => patternMatch(s, expr))) {
+        scopes.push(expr);
+        return true;
+      }
+      return false;
+    }
+
+    if (expr.hasOwnProperty('AllOf')) {
+      let scopeIndex = scopes.length;
```

Comment:
`startingIndex`?

Not necessary to change, just makes it easier to understand for me.